app/jobs/google/distribute_search_job.rb

```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+module Google
+  class DistributeSearchJob < ApplicationJob
+    queue_as :default
+
+    def perform(keyword_ids)
+      keyword_ids.each do |keyword_id|
```

Comment:
> You need to find a way to work around the limitations of mass-searching keywords as Google prevents it. Be creative.

I think to handle this requirement we need to add some interval between requests as If we do massive requests at the same time, Google blocks the IP. Adding some interval between  request/jobs works.(`perform_later` start the job immediately)
```ruby
     keyword_ids.each_with_index do |keyword_id, index|
        SearchKeywordJob.set(wait: index + 3).perform_later(keyword_id)
     end
```

Then it will enqueue the jobs immediately, but start the job based on set time, after the first job running, 2nd job will run 3 seconds later after running the first job, the third job will run 3 seconds later after running the 2nd job, and so on. It will prevent Google blocking.

What do you think?

Reply:
Interesting @junan  :) 

I have performed some more tests, it seems that there are several threads running concurrently (4 on my local, 10 by default according to google?). My tests shown that the set `wait:` time is not relative to previous tasks, but it's an absolute time **based from the enqueued time**. So I would prefer setting `index * 3`, so that each t+1 tasks will have an additional 3 seconds waiting time. 

Also, failed jobs will currently be retried every 5 seconds. Considering 4 simultaneous tasks, it's approx 1.2s of wait between retries. We should expect to wait 'more' when retrying (because 99% of the time, the reason why it would failed will be `too many requests`). What about setting 4 tasks * 3 seconds = 12 seconds for retrial time too? :)
Thanks,

Reply:
Update: Even 3 seconds seems not enough.
I'm not sure exactly that is the mechanism that bring them to start all together by bach of 4... (4 = number of threads available). 

I will perform more tests with longer wait time.

```bash
00 2021-07-19T06:53:21.043Z pid=70693 tid=35 class=Google::DistributeSearchJob jid=bd72b4adf67169d7fb6f58d2 elapsed=0.393 INFO: done
01 2021-07-19T06:53:22.466Z pid=70693 tid=5t  class=Google::SearchKeywordJob jid=1 INFO: start
02 2021-07-19T06:53:23.645Z pid=70693 tid=5t  class=Google::SearchKeywordJob jid=1 elapsed=1.179 INFO: done
04 2021-07-19T06:53:25.014Z pid=70693 tid=2kp class=Google::SearchKeywordJob jid=2 INFO: start
05 2021-07-19T06:53:26.157Z pid=70693 tid=2kp class=Google::SearchKeywordJob jid=2 elapsed=1.143 INFO: done
09 2021-07-19T06:53:30.411Z pid=70693 tid=2o1 class=Google::SearchKeywordJob jid=3 INFO: start
09 2021-07-19T06:53:30.417Z pid=70693 tid=35  class=Google::SearchKeywordJob jid=4 INFO: start
10 2021-07-19T06:53:31.312Z pid=70693 tid=35  class=Google::SearchKeywordJob jid=4 elapsed=0.895 INFO: done
10 2021-07-19T06:53:31.427Z pid=70693 tid=2o1 class=Google::SearchKeywordJob jid=3 elapsed=1.016 INFO: done
13 2021-07-19T06:53:34.623Z pid=70693 tid=35  class=Google::SearchKeywordJob jid=5 INFO: start
14 2021-07-19T06:53:35.558Z pid=70693 tid=35  class=Google::SearchKeywordJob jid=5 elapsed=0.935 INFO: done
16 2021-07-19T06:53:37.763Z pid=70693 tid=5t  class=Google::SearchKeywordJob jid=6 INFO: start
16 2021-07-19T06:53:37.770Z pid=70693 tid=2kp class=Google::SearchKeywordJob jid=7 INFO: start
17 2021-07-19T06:53:38.757Z pid=70693 tid=2kp class=Google::SearchKeywordJob jid=7 elapsed=0.987 INFO: done
17 2021-07-19T06:53:38.769Z pid=70693 tid=5t  class=Google::SearchKeywordJob jid=6 elapsed=1.005 INFO: done
20 2021-07-19T06:53:41.459Z pid=70693 tid=35  class=Google::SearchKeywordJob jid=8 INFO: start
20 2021-07-19T06:53:41.465Z pid=70693 tid=2o1 class=Google::SearchKeywordJob jid=9 INFO: start
21 2021-07-19T06:53:42.964Z pid=70693 tid=2o1 class=Google::SearchKeywordJob jid=9 elapsed=1.499 INFO: done
22 2021-07-19T06:53:43.895Z pid=70693 tid=35  class=Google::SearchKeywordJob jid=8 elapsed=2.436 INFO: done
26 2021-07-19T06:53:47.571Z pid=70693 tid=35  class=Google::SearchKeywordJob jid=10 INFO: start
26 2021-07-19T06:53:47.578Z pid=70693 tid=5t  class=Google::SearchKeywordJob jid=11 INFO: start
26 21021-07-19T06:53:47.584Z pid=70693 tid=2kp class=Google::SearchKeywordJob jid=12 INFO: start
26 2021-07-19T06:53:47.590Z pid=70693 tid=2o1 class=Google::SearchKeywordJob jid=13 INFO: start
```

Reply:
(not that I carefully flush the queues between 2 tests, to ensure there is no previous job that will interfere)

Reply:
> So I would prefer setting index * 3, so that each t+1 tasks will have an additional 3 seconds waiting time.
> What about setting 4 tasks * 3 seconds = 12 seconds for retrial time too? :)

Sure, we can set more interval time based on our needs.

Reply:
I forgot to mentionned that on the left side of the log, you have the cumulated seconds.
26 seconds total wait for the 13th job is much lower than 13*3seconds = 39seconds. So the wait command is weird... humhum, any idea?

Reply:
Okay... more research simply indicates that sidekiq is not a second-grade precision tool ^^
I will then keep those values : 3s between each search (as you suggested) and 12 seconds on retrial). 
Thanks @Junan for this interesting feedback :)

Reply:
Done~